extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmCdiProcessor.java

```diff
@@ -140,28 +144,46 @@ void generateJpaConfigBean(HibernateOrmRecorder recorder,
                 .scope(Singleton.class)
                 .unremovable()
                 .setRuntimeInit()
-                .supplier(recorder.jpaConfigSupplier(hibernateOrmRuntimeConfig))
-                .destroyer(JPAConfig.Destroyer.class);
-
-        // Add a synthetic dependency from JPAConfig to any datasource/pool,
-        // so that JPAConfig is destroyed before the datasource/pool.
-        // The alternative would be adding an application destruction observer
-        // (@Observes @BeforeDestroyed(ApplicationScoped.class)) to JPAConfig,
-        // but that would force initialization of JPAConfig upon application shutdown,
-        // which may cause cascading failures if the shutdown happened before JPAConfig was initialized.
-        if (capabilities.isPresent(Capability.HIBERNATE_REACTIVE)) {
-            configurator.addInjectionPoint(ParameterizedType.create(DotName.createSimple(Instance.class),
-                    new Type[] { ClassType.create(DotName.createSimple("io.vertx.sqlclient.Pool")) }, null),
-                    AnnotationInstance.builder(Any.class).build());
-        } else {
-            configurator.addInjectionPoint(ParameterizedType.create(DotName.createSimple(Instance.class),
-                    new Type[] { ClassType.create(DotName.createSimple(AgroalDataSource.class)) }, null),
-                    AnnotationInstance.builder(Any.class).build());
-        }
+                .supplier(recorder.jpaConfigSupplier(hibernateOrmRuntimeConfig));
 
         syntheticBeanBuildItemBuildProducer.produce(configurator.done());
     }
 
+    @BuildStep
+    @Record(ExecutionTime.RUNTIME_INIT)
+    void generateJpaConfigBeanObserver(
+            HibernateOrmRecorder recorder,
+            ObserverRegistrationPhaseBuildItem observerRegistrationPhase,
+            BuildProducer<ObserverRegistrationPhaseBuildItem.ObserverConfiguratorBuildItem> observerConfigurationRegistry) {
+        observerConfigurationRegistry.produce(
+                new ObserverRegistrationPhaseBuildItem.ObserverConfiguratorBuildItem(observerRegistrationPhase.getContext()
+                        .configure()
+                        .beanClass(DotName.createSimple("io.quarkus.hibernate.orm.runtime.JPAConfig"))
+                        .observedType(ShutdownEvent.class)
+                        .priority(JPA_CONFIG_SHUTDOWN_PRIORITY)
+                        .notify(mc -> {
+                            // Essentially do the following:
+                            // Arc.container().instance( JPAConfig.class ).get().shutdown();
```

Comment:
Can't this be simply written in Java in a recorder method?

Reply:
In theory, you could copy the logic from `LifecycleEventsBuildStep#startupEvent()`, i.e. consume the `BeanContainerBuildItem` and `ShutdownContextBuildItem`, and register a shutdown task in the recorder. This works because tasks are executed in the reverse order and since you consume the `BeanContainerBuildItem` the shutdown task is registered after the task that shuts down the CDI container.

Reply:
another idea I had here was to put 
```java
Arc.container().instance( JPAConfig.class ).get().shutdown();
```
inside a static method of `JPAConfig`, and then, this handler would make a single call to invoke a static method (that is if we want to have fewer calls to various `mc` methods).

Reply:
It's more that I'd avoid hard-to-read bytecode generation if I can, but if this particular feature isn't compatible with recorders... so be it.
Thanks for looking into it.

Reply:
> another idea I had here was to put
> 
> ```java
> Arc.container().instance( JPAConfig.class ).get().shutdown();
> ```
> 
> inside a static method of `JPAConfig`, and then, this handler would make a single call to invoke a static method (that is if we want to have fewer calls to various `mc` methods).

:+1: That's also doable and often used.